File: movies.py
import csv
from collections import defaultdict
import time
import random

class Movie:
    def __init__(self, title : str, genre : str, year : int):     
        self.title = title
        self.genre = genre
        self.year = year
    def __repr__(self):
        return f'({self.title} : {self.year} : {self.genre})'

movie_list = [] # list of movie objects, assuming id=index
movie_dic = defaultdict(set)  # dictionary of sets, keyed on tuple(year, genre)
unique_years = set()
unique_genres = set()


def load():
    col_title = 1
    col_genre = 2
    col_year = 6
    with open("IMDB-Movie-Data.csv", "r") as f:
        reader = csv.reader(f, delimiter=",")
        for i,row in enumerate(reader):
            if i == 0:            
                col_title = row.index('Title')
                col_genre = row.index('Genre')
                col_year = row.index('Year')
            else:
                #print (row)
                title = row[col_title]
                genre = row[col_genre]
                year = int(row[col_year])
                id = i-1
                #print(f' id: {id}, title: {title} genre: {genre} year: {year}')

                movie_list.append(Movie(title, genre, year))

                unique_years.add(year)

                genres = genre.split(',')
                for g in genres:
                    key = (year,g.casefold())
                    movie_dic[key].add(id)

                    unique_genres.add(g)

    #print(sorted(unique_years))
    #print(sorted(unique_genres))

load()


def get_uniques(genre: str = '', from_year: int = 0, to_year: int = 3000):
    ids = set()
    year_query = set(range(from_year, to_year+1)).intersection(unique_years)
    genre_query = (genre,) if genre != '' else unique_genres
    for y in year_query:
            for g in genre_query:
                key = (y, g.casefold())
                if key in movie_dic.keys():
                    ids.update(movie_dic[key])

    return [movie_list[x] for x in ids]


# retrieve movies in genre 'Animation', any date
res = get_uniques('animation')
print('### animation, any date ###')
print(res)

# retrieve movies in genre 'Action', 2006-2007
res = get_uniques('action', 2006, 2007)
print('### action, 2006-2007 ##')
print(res)

# retrieve movies for any genre, 2006
res = get_uniques('', 2006, 2006)
print('### any genre, 2006 ##')
print(res)


# test average retrieval time over 1000 random queries
def test_speed():
    trials = 1000*100
    tot_secs = 0
    for i in range(0, trials):
        genre = random.choice(list(unique_genres))
        year_from = random.choice(list(unique_years))
        year_to = random.choice([y for y in list(unique_years) if y >= year_from])
        tic =  time.perf_counter()
        movies = get_uniques(genre, year_from, year_to)
        toc = time.perf_counter()    
        tot_secs += (toc - tic)
    avg_secs = tot_secs / trials
    print(f'avg {avg_secs*1000} ms')

test_speed()

File: movies.cs
using System.Diagnostics;
using System.Text.RegularExpressions;

struct Movie
{
    public int Id { get; set; }
    public string Title { get; set; }
    public int Year { get; set; }
    public string Genres { get; set; }

    public override string ToString() => $"{Title} : {Year} : ({Genres})";
}

class Movies
{
    List<Movie> movieList = new List<Movie>();
    Dictionary<(int, string), HashSet<int>> movieDict = new Dictionary<(int, string), HashSet<int>>();
    HashSet<int> uniqueYears = new HashSet<int>();
    HashSet<string> uniqueGenres = new HashSet<string>();

    static void Main(string[] args)
    {
        var m = new Movies();
        m.Load();

        List<Movie> m1 = m.GetUniques("Animation", 2015, 2016);
        //m1.ForEach(x => Console.WriteLine(x));
        Console.WriteLine(string.Join(", ", m1));

        List<Movie> m2 = m.GetUniques("", 2012, 2012);
        //m2.ForEach(x => Console.WriteLine(x));
        Console.WriteLine(string.Join(", ", m2));

        m.TestSpeed();
    }

    public void Load()
    {
        using (var reader = new StreamReader("IMDB-Movie-Data.csv"))
        {
            var colIndices = new Dictionary<string, int>();
            int lineIndex = 0;
            while(!reader.EndOfStream)
            {
                string line = reader.ReadLine();
                // string[] values = line.Split(",");
                // ignore commas inside quotes and remove escaped quotes
                string[] values = Regex.Split(line, "[,]{1}(?=(?:[^\"]*\"[^\"]*\")*(?![^\"]*\"))").Select(x => x.Replace("\"", "")).ToArray();

                if (lineIndex == 0)
                {
                    // header row                   
                    colIndices = Enumerable.Range(0, values.Length).ToDictionary(x => values[x]); // indices are the values, values[x] are the keys
                }
                else
                {
                    // data row
                    int id = lineIndex - 1;
                    string title = values[colIndices["Title"]];
                    int year = int.Parse(values[colIndices["Year"]]);
                    string genres = values[colIndices["Genre"]];

                    var m = new Movie { Id = id, Title = title, Year = year, Genres = genres };
                    movieList.Add(m);

                    uniqueYears.Add(year);

                    foreach (string g in genres.Split(","))
                    {
                        (int, string) key = (year, g);

                        if (!movieDict.ContainsKey(key)) movieDict.Add(key, new HashSet<int>());
                        movieDict[key].Add(id);

                        uniqueGenres.Add(g);
                    }
                }

                lineIndex++;
            }
        }
    }

    public List<Movie> GetUniques(string genre, int fromYear, int toYear)
    {
        var uniques = new HashSet<int>();
        
        HashSet<int> yearQuery = new HashSet<int>(Enumerable.Range(fromYear, toYear+1-fromYear)); 
        yearQuery.IntersectWith(uniqueYears); // returns void, cannot chain

        List<string> genreQuery = String.IsNullOrEmpty(genre) ? new List<string>(uniqueGenres) : [genre];

        foreach (int y in yearQuery)
        {
            foreach(string g in genreQuery)
            {
                (int, string) key = (y, g);
                if (movieDict.TryGetValue(key, out var indices))
                    uniques.UnionWith(indices);
            }
        }

        return uniques.Select(x => movieList[x]).ToList();
    }

    public void TestSpeed()
    {
        var rnd = new Random();
        int trials = 100_000;
        double totalMs = 0;
        foreach (int i in Enumerable.Range(0, trials))
        {
            string genre = rnd.GetItems(uniqueGenres.ToArray(), 1).First();
            int yearFrom = rnd.GetItems(uniqueYears.ToArray(), 1).First();
            int yearTo = rnd.GetItems(uniqueYears.Where(x => x >= yearFrom).ToArray(), 1).First();

            Stopwatch s = Stopwatch.StartNew();
            List<Movie> m = GetUniques(genre, yearFrom, yearTo);            
            totalMs += s.Elapsed.TotalMilliseconds;
        }
        double avgMs = totalMs / trials;
        Console.WriteLine($"avg {avgMs} ms");
    }
}
